File: git-worktree-manager-intellij/src/main/kotlin/com/iambipinpaul/gitworktreemanager/services/GitService.kt
package com.iambipinpaul.gitworktreemanager.services

import com.iambipinpaul.gitworktreemanager.models.GitCommandResult
import com.iambipinpaul.gitworktreemanager.models.Worktree
import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.CapturingProcessHandler
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.SystemInfo
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

data class WorktreeStatus(
    val modifiedCount: Int,
    val untrackedCount: Int,
    val incoming: Int,
    val outgoing: Int,
    val hasUpstream: Boolean = true,
)

class GitService(
    private val logger: Logger = Logger.getInstance(GitService::class.java),
) {
    suspend fun getWorktrees(repositoryPath: String): GitCommandResult<List<Worktree>> {
        logger.info("Getting worktrees for repository: $repositoryPath")

        val result = executeGitCommand(repositoryPath, listOf("worktree", "list", "--porcelain"))
        if (!result.success) {
            logger.error("Failed to list worktrees: ${result.errorMessage}")
            return GitCommandResult.fail(result.errorMessage ?: "Failed to list worktrees", result.exitCode)
        }

        val worktrees = WorktreeParser.parsePorcelainOutput(result.output ?: "")
        logger.info("Found ${worktrees.size} worktree(s)")
        return GitCommandResult.ok(worktrees)
    }

    suspend fun addWorktree(
        repositoryPath: String,
        worktreePath: String,
        branchName: String,
        createBranch: Boolean = false,
        baseBranch: String? = null,
    ): GitCommandResult<Unit> {
        val args = mutableListOf("worktree", "add")
        if (createBranch) {
            args.addAll(listOf("-b", branchName))
            args.add(worktreePath)
            if (!baseBranch.isNullOrBlank()) {
                args.add(baseBranch)
            }
        } else {
            args.add(worktreePath)
            args.add(branchName)
        }

        logger.info(
            "Adding worktree: path='$worktreePath', branch='$branchName', " +
                "createBranch=$createBranch, baseBranch='$baseBranch'",
        )

        val result = executeGitCommand(repositoryPath, args)
        return if (result.success) {
            logger.info("Successfully added worktree at '$worktreePath'")
            GitCommandResult.ok()
        } else {
            logger.error("Failed to add worktree: ${result.errorMessage}")
            GitCommandResult.fail(result.errorMessage ?: "Failed to add worktree", result.exitCode)
        }
    }

    suspend fun removeWorktree(
        repositoryPath: String,
        worktreePath: String,
        force: Boolean = false,
    ): GitCommandResult<Unit> {
        // A single --force removes dirty worktrees, but locked worktrees require
        // -f -f, so send --force twice whenever force is requested.
        val args = mutableListOf("worktree", "remove")
        if (force) {
            args.add("--force")
            args.add("--force")
        }
        args.add(worktreePath)

        logger.info("Removing worktree: path='$worktreePath', force=$force")

        val result = executeGitCommand(repositoryPath, args)
        return if (result.success) {
            logger.info("Successfully removed worktree at '$worktreePath'")
            GitCommandResult.ok()
        } else {
            logger.error("Failed to remove worktree: ${result.errorMessage}")
            GitCommandResult.fail(result.errorMessage ?: "Failed to remove worktree", result.exitCode)
        }
    }

    suspend fun getRepositoryRoot(path: String): String? {
        logger.info("Getting repository root for path: $path")

        val result = executeGitCommand(path, listOf("rev-parse", "--show-toplevel"))
        if (!result.success || result.output.isNullOrBlank()) {
            logger.warn("Path '$path' is not within a Git repository")
            return null
        }

        val root = result.output.trim()
        logger.info("Repository root: $root")
        return root
    }

    suspend fun isGitInstalled(): Boolean {
        logger.info("Checking if Git is installed")

        return try {
            val result = executeGitCommand(File(".").absolutePath, listOf("--version"))
            if (result.success) {
                logger.info("Git is installed: ${result.output?.trim()}")
            } else {
                logger.warn("Git is not installed or not in PATH")
            }
            result.success
        } catch (ex: Exception) {
            logger.error("Error checking Git installation", ex)
            false
        }
    }

    suspend fun getBranches(repositoryPath: String): GitCommandResult<List<String>> {
        logger.info("Getting branches for repository: $repositoryPath")

        val localResult = executeGitCommand(
            repositoryPath,
            listOf("branch", "--format=%(refname:short)"),
        )
        val remoteResult = executeGitCommand(
            repositoryPath,
            listOf("branch", "-r", "--format=%(refname:short)"),
        )

        val branches = mutableListOf<String>()

        if (localResult.success && !localResult.output.isNullOrBlank()) {
            val localBranches = localResult.output
                .lineSequence()
                .map { it.trim() }
                .filter { it.isNotEmpty() }
            branches.addAll(localBranches)
        }

        if (remoteResult.success && !remoteResult.output.isNullOrBlank()) {
            val remoteBranches = remoteResult.output
                .lineSequence()
                .map { it.trim() }
                .filter { it.isNotEmpty() && !it.contains("HEAD") }
                .map { if (it.startsWith("origin/")) it.substring(7) else it }
                .filter { it !in branches }
            branches.addAll(remoteBranches)
        }

        logger.info("Found ${branches.size} branch(es)")
        return GitCommandResult.ok(branches.distinct().sorted())
    }

    suspend fun getWorktreeStatus(path: String, branch: String): WorktreeStatus {
        try {
            if (path.isBlank() || !File(path).exists()) {
                return WorktreeStatus(0, 0, 0, 0, false)
            }

            val statusResult = executeGitCommand(
                path,
                listOf("status", "--porcelain=v1", "-unormal"),
                timeoutMs = STATUS_TIMEOUT_MS,
            )

            var modified = 0
            var untracked = 0

            if (statusResult.success && !statusResult.output.isNullOrBlank()) {
                for (line in statusResult.output.lineSequence()) {
                    if (line.length < 3) {
                        continue
                    }
                    val status = line.substring(0, 2)
                    if (status == "??") {
                        untracked++
                    } else {
                        modified++
                    }
                }
            }

            var ahead = 0
            var behind = 0
            var hasUpstream = false

            try {
                val upstreamResult = executeGitCommand(
                    path,
                    listOf("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}"),
                    timeoutMs = UPSTREAM_TIMEOUT_MS,
                )

                if (upstreamResult.success && !upstreamResult.output.isNullOrBlank()) {
                    hasUpstream = true
                    val countResult = executeGitCommand(
                        path,
                        listOf("rev-list", "--left-right", "--count", "HEAD...@{u}"),
                        timeoutMs = UPSTREAM_TIMEOUT_MS,
                    )

                    if (countResult.success && !countResult.output.isNullOrBlank()) {
                        val parts = countResult.output.trim()
                            .split(Regex("\\s+"))
                            .filter { it.isNotBlank() }
                        if (parts.size >= 2) {
                            ahead = parts[0].toIntOrNull() ?: 0
                            behind = parts[1].toIntOrNull() ?: 0
                        }
                    }
                }
            } catch (ex: Exception) {
                logger.warn("No upstream found for $path. Branch is local-only.")
                hasUpstream = false
            }

            return WorktreeStatus(modified, untracked, behind, ahead, hasUpstream)
        } catch (ex: CancellationException) {
            logger.warn("Status check cancelled for worktree: $path (branch=$branch)")
            return WorktreeStatus(0, 0, 0, 0, false)
        } catch (ex: Exception) {
            logger.error("Failed to get status for $path", ex)
            return WorktreeStatus(0, 0, 0, 0, false)
        }
    }

    private suspend fun executeGitCommand(
        workingDirectory: String,
        arguments: List<String>,
        timeoutMs: Int = DEFAULT_TIMEOUT_MS,
    ): GitProcessResult {
        val enableLongPathSupport = SystemInfo.isWindows
        val gitArguments = buildGitArguments(arguments, enableLongPathSupport)

        logger.info("Executing: git ${gitArguments.joinToString(" ")} (in $workingDirectory)")

        val commandLine = GeneralCommandLine(GIT_EXECUTABLE)
            .withWorkDirectory(workingDirectory)
            .withCharset(Charsets.UTF_8)
            .withParameters(gitArguments)

        val handler = CapturingProcessHandler(commandLine)

        return withContext(Dispatchers.IO) {
            try {
                val output = handler.runProcess(timeoutMs)
                if (output.isTimeout) {
                    handler.destroyProcess()
                    logger.error("Git command timed out after ${timeoutMs}ms: git ${gitArguments.joinToString(" ")}")
                    return@withContext GitProcessResult(
                        success = false,
                        exitCode = -1,
                        output = null,
                        errorMessage = "Git command timed out",
                    )
                }

                val exitCode = output.exitCode
                val stdout = output.stdout
                val stderr = output.stderr

                if (stdout.isNotBlank()) {
                    logger.info("Git stdout:\n${stdout.trim()}")
                }
                if (stderr.isNotBlank()) {
                    if (exitCode == 0) {
                        logger.warn("Git stderr (exit code 0):\n${stderr.trim()}")
                    } else {
                        logger.error("Git stderr (exit code $exitCode):\n${stderr.trim()}")
                    }
                }

                logger.info("Git command completed with exit code: $exitCode")

                GitProcessResult(
                    success = exitCode == 0,
                    exitCode = exitCode,
                    output = stdout,
                    errorMessage = if (exitCode != 0) {
                        createUserFacingErrorMessage(
                            stderr.trim().ifBlank { "Git command failed with exit code $exitCode." },
                            enableLongPathSupport,
                        )
                    } else {
                        null
                    },
                )
            } catch (ex: CancellationException) {
                handler.destroyProcess()
                logger.warn("Git command was cancelled: git ${gitArguments.joinToString(" ")}")
                GitProcessResult(
                    success = false,
                    exitCode = -1,
                    output = null,
                    errorMessage = "Git command was cancelled",
                )
            } catch (ex: Exception) {
                logger.error("Failed to execute Git command: git ${gitArguments.joinToString(" ")}", ex)
                GitProcessResult(
                    success = false,
                    exitCode = -1,
                    output = null,
                    errorMessage = createUserFacingErrorMessage(
                        "Failed to execute Git command: ${ex.message}",
                        enableLongPathSupport,
                    ),
                )
            }
        }
    }

    private data class GitProcessResult(
        val success: Boolean,
        val exitCode: Int,
        val output: String?,
        val errorMessage: String?,
    )

    companion object {
        private const val DEFAULT_TIMEOUT_MS = 30_000
        private const val STATUS_TIMEOUT_MS = 10_000
        private const val UPSTREAM_TIMEOUT_MS = 5_000
        private const val GIT_EXECUTABLE = "git"
        private val LONG_PATH_ERROR_PATTERNS = listOf(
            "filename too long",
            "file name too long",
            "path too long",
            "path length",
            "path, file name, or both are too long",
        )

        internal fun buildGitArguments(
            arguments: List<String>,
            enableLongPathSupport: Boolean,
        ): List<String> {
            return if (enableLongPathSupport) {
                listOf("-c", "core.longpaths=true") + arguments
            } else {
                arguments
            }
        }

        internal fun isLongPathError(errorMessage: String?): Boolean {
            if (errorMessage.isNullOrBlank()) {
                return false
            }

            return LONG_PATH_ERROR_PATTERNS.any { pattern ->
                errorMessage.contains(pattern, ignoreCase = true)
            }
        }

        internal fun createUserFacingErrorMessage(
            errorMessage: String,
            longPathSupportWasEnabled: Boolean,
        ): String {
            if (!isLongPathError(errorMessage)) {
                return errorMessage
            }

            val supportStatus = if (longPathSupportWasEnabled) {
                "Git Worktree Hub already ran this Git command with process-scoped core.longpaths=true."
            } else {
                "This looks like a Git for Windows long path error."
            }

            return listOf(
                errorMessage,
                supportStatus +
                    " If the same repository also fails from a terminal or another Git tool, " +
                    "enable Git for Windows long paths globally with:",
                "git config --global core.longpaths true",
                "This is a Git for Windows setting, not a JetBrains IDE setting.",
            ).joinToString(System.lineSeparator() + System.lineSeparator())
        }
    }
}

File: GitWorktreeManager.Core/Services/GitService.cs
namespace GitWorktreeManager.Services;

using System.Diagnostics;
using System.Text;
using Models;

/// <summary>
/// Service for executing Git worktree commands using the Git CLI.
/// </summary>
public class GitService : IGitService
{
    private const int DefaultTimeoutMs = 30000; // 30 seconds
    private const string GitExecutable = "git";
    private static readonly string[] LongPathErrorPatterns =
    {
        "filename too long",
        "file name too long",
        "path too long",
        "path length",
        "path, file name, or both are too long"
    };

    private readonly ILoggerService? _logger;

    /// <summary>
    /// Initializes a new instance of the <see cref="GitService"/> class.
    /// </summary>
    public GitService() : this(null)
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="GitService"/> class with logging support.
    /// </summary>
    /// <param name="logger">The logger service for recording operations.</param>
    public GitService(ILoggerService? logger)
    {
        _logger = logger;
    }

    /// <inheritdoc />
    public async Task<GitCommandResult<IReadOnlyList<Worktree>>> GetWorktreesAsync(
        string repositoryPath,
        CancellationToken cancellationToken = default)
    {
        _logger?.LogInformation($"Getting worktrees for repository: {repositoryPath}");

        GitProcessResult result = await ExecuteGitCommandAsync(
            repositoryPath,
            new[] { "worktree", "list", "--porcelain" },
            cancellationToken);

        if (!result.Success)
        {
            _logger?.LogError($"Failed to list worktrees: {result.ErrorMessage}");
            return GitCommandResult<IReadOnlyList<Worktree>>.Fail(
                result.ErrorMessage ?? "Failed to list worktrees",
                result.ExitCode);
        }

        IReadOnlyList<Worktree> worktrees = WorktreeParser.ParsePorcelainOutput(result.Output ?? string.Empty);
        _logger?.LogInformation($"Found {worktrees.Count} worktree(s)");
        return GitCommandResult<IReadOnlyList<Worktree>>.Ok(worktrees);
    }

    /// <inheritdoc />
    public async Task<GitCommandResult> AddWorktreeAsync(
        string repositoryPath,
        string worktreePath,
        string branchName,
        bool createBranch = false,
        string? baseBranch = null,
        CancellationToken cancellationToken = default)
    {
        // Build command: git worktree add [-b <branch>] <path> [<commit-ish>]
        // With -b: git worktree add -b <new-branch> <path> [<base-branch>]
        // Without -b: git worktree add <path> <branch>
        // Passed as an argument list (no shell quoting) so paths/branches with
        // spaces or quotes are handled safely.
        List<string> arguments;

        if (createBranch)
        {
            // Create new branch based on another branch
            if (!string.IsNullOrEmpty(baseBranch))
            {
                arguments = new List<string> { "worktree", "add", "-b", branchName, worktreePath, baseBranch };
            }
            else
            {
                arguments = new List<string> { "worktree", "add", "-b", branchName, worktreePath };
            }
        }
        else
        {
            // Checkout existing branch
            arguments = new List<string> { "worktree", "add", worktreePath, branchName };
        }

        _logger?.LogInformation(
            $"Adding worktree: path='{worktreePath}', branch='{branchName}', createBranch={createBranch}, baseBranch='{baseBranch}'");

        GitProcessResult result = await ExecuteGitCommandAsync(
            repositoryPath,
            arguments,
            cancellationToken);

        if (result.Success)
        {
            _logger?.LogInformation($"Successfully added worktree at '{worktreePath}'");
            return GitCommandResult.Ok();
        }
        else
        {
            _logger?.LogError($"Failed to add worktree: {result.ErrorMessage}");
            return GitCommandResult.Fail(result.ErrorMessage ?? "Failed to add worktree", result.ExitCode);
        }
    }

    /// <inheritdoc />
    public async Task<GitCommandResult> RemoveWorktreeAsync(
        string repositoryPath,
        string worktreePath,
        bool force = false,
        CancellationToken cancellationToken = default)
    {
        // Build command: git worktree remove [--force] <path>
        // A single --force removes dirty worktrees, but locked worktrees require
        // -f -f (e.g. "cannot remove a locked working tree ... use 'remove -f -f'").
        // Send --force twice whenever force is requested so one flag covers both.
        List<string> arguments = force
            ? new List<string> { "worktree", "remove", "--force", "--force", worktreePath }
            : new List<string> { "worktree", "remove", worktreePath };

        _logger?.LogInformation($"Removing worktree: path='{worktreePath}', force={force}");

        GitProcessResult result = await ExecuteGitCommandAsync(
            repositoryPath,
            arguments,
            cancellationToken);

        if (result.Success)
        {
            _logger?.LogInformation($"Successfully removed worktree at '{worktreePath}'");
            return GitCommandResult.Ok();
        }
        else
        {
            _logger?.LogError($"Failed to remove worktree: {result.ErrorMessage}");
            return GitCommandResult.Fail(result.ErrorMessage ?? "Failed to remove worktree", result.ExitCode);
        }
    }

    /// <inheritdoc />
    public async Task<string?> GetRepositoryRootAsync(
        string path,
        CancellationToken cancellationToken = default)
    {
        _logger?.LogInformation($"Getting repository root for path: {path}");

        GitProcessResult result = await ExecuteGitCommandAsync(
            path,
            new[] { "rev-parse", "--show-toplevel" },
            cancellationToken);

        if (!result.Success || string.IsNullOrWhiteSpace(result.Output))
        {
            _logger?.LogWarning($"Path '{path}' is not within a Git repository");
            return null;
        }

        string root = result.Output.Trim();
        _logger?.LogInformation($"Repository root: {root}");
        return root;
    }

    /// <inheritdoc />
    public async Task<bool> IsGitInstalledAsync(
        CancellationToken cancellationToken = default)
    {
        _logger?.LogInformation("Checking if Git is installed");

        try
        {
            GitProcessResult result = await ExecuteGitCommandAsync(
                Directory.GetCurrentDirectory(),
                new[] { "--version" },
                cancellationToken);

            if (result.Success)
            {
                _logger?.LogInformation($"Git is installed: {result.Output?.Trim()}");
            }
            else
            {
                _logger?.LogWarning("Git is not installed or not in PATH");
            }

            return result.Success;
        }
        catch (Exception ex)
        {
            _logger?.LogException(ex, "Error checking Git installation");
            return false;
        }
    }

    /// <inheritdoc />
    public async Task<GitCommandResult<IReadOnlyList<string>>> GetBranchesAsync(
        string repositoryPath,
        CancellationToken cancellationToken = default)
    {
        _logger?.LogInformation($"Getting branches for repository: {repositoryPath}");

        // Get local branches
        GitProcessResult localResult = await ExecuteGitCommandAsync(
            repositoryPath,
            new[] { "branch", "--format=%(refname:short)" },
            cancellationToken);

        // Get remote branches
        GitProcessResult remoteResult = await ExecuteGitCommandAsync(
            repositoryPath,
            new[] { "branch", "-r", "--format=%(refname:short)" },
            cancellationToken);

        var branches = new List<string>();

        if (localResult.Success && !string.IsNullOrWhiteSpace(localResult.Output))
        {
            IEnumerable<string> localBranches = localResult.Output
                .Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(b => b.Trim())
                .Where(b => !string.IsNullOrEmpty(b));
            branches.AddRange(localBranches);
        }

        if (remoteResult.Success && !string.IsNullOrWhiteSpace(remoteResult.Output))
        {
            IEnumerable<string> remoteBranches = remoteResult.Output
                .Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(b => b.Trim())
                .Where(b => !string.IsNullOrEmpty(b) && !b.Contains("HEAD"))
                // Remove origin/ prefix for cleaner display but keep track it's remote
                .Select(b => b.StartsWith("origin/") ? b.Substring(7) : b)
                .Where(b => !branches.Contains(b)); // Don't duplicate local branches
            branches.AddRange(remoteBranches);
        }

        _logger?.LogInformation($"Found {branches.Count} branch(es)");
        return GitCommandResult<IReadOnlyList<string>>.Ok(branches.Distinct().OrderBy(b => b).ToList());
    }


    /// <summary>
    /// Executes a Git command and returns the result.
    /// Arguments are passed via <see cref="ProcessStartInfo.ArgumentList"/> (no shell),
    /// so paths/branches with spaces or quotes are handled safely.
    /// </summary>
    /// <param name="workingDirectory">The working directory for the command.</param>
    /// <param name="arguments">The Git command arguments (already split).</param>
    /// <param name="cancellationToken">Cancellation token for the operation.</param>
    /// <returns>The result of the command execution.</returns>
    private async Task<GitProcessResult> ExecuteGitCommandAsync(
        string workingDirectory,
        IReadOnlyList<string> arguments,
        CancellationToken cancellationToken)
    {
        bool enableLongPathSupport = OperatingSystem.IsWindows();
        List<string> fullArgs = BuildGitArgumentList(arguments, enableLongPathSupport);
        string displayArgs = string.Join(" ", fullArgs.Select(QuoteForDisplay));

        _logger?.LogInformation($"Executing: git {displayArgs} (in {workingDirectory})");

        var startInfo = new ProcessStartInfo
        {
            FileName = GitExecutable,
            WorkingDirectory = workingDirectory,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
            StandardOutputEncoding = Encoding.UTF8,
            StandardErrorEncoding = Encoding.UTF8
        };
        foreach (string arg in fullArgs)
        {
            startInfo.ArgumentList.Add(arg);
        }

        using var process = new Process { StartInfo = startInfo };
        var stdout = new StringBuilder();
        var stderr = new StringBuilder();

        process.OutputDataReceived += (_, e) =>
        {
            if (e.Data != null)
            {
                stdout.AppendLine(e.Data);
            }
        };

        process.ErrorDataReceived += (_, e) =>
        {
            if (e.Data != null)
            {
                stderr.AppendLine(e.Data);
            }
        };

        try
        {
            process.Start();
            process.BeginOutputReadLine();
            process.BeginErrorReadLine();

            // Wait for process with timeout and cancellation support
            using var timeoutCts = new CancellationTokenSource(DefaultTimeoutMs);
            using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(
                cancellationToken, timeoutCts.Token);

            try
            {
                await process.WaitForExitAsync(linkedCts.Token);
            }
            catch (OperationCanceledException) when (timeoutCts.IsCancellationRequested)
            {
                TryKillProcess(process);
                _logger?.LogError($"Git command timed out after {DefaultTimeoutMs}ms: git {displayArgs}");
                return new GitProcessResult { Success = false, ExitCode = -1, ErrorMessage = "Git command timed out" };
            }
            catch (OperationCanceledException)
            {
                TryKillProcess(process);
                _logger?.LogWarning($"Git command was cancelled: git {displayArgs}");
                return new GitProcessResult
                {
                    Success = false, ExitCode = -1, ErrorMessage = "Git command was cancelled"
                };
            }

            int exitCode = process.ExitCode;
            string output = stdout.ToString();
            string error = stderr.ToString();

            // Log stdout if present
            if (!string.IsNullOrWhiteSpace(output))
            {
                _logger?.LogInformation($"Git stdout:\n{output.Trim()}");
            }

            // Log stderr if present (as warning for exit code 0, error otherwise)
            if (!string.IsNullOrWhiteSpace(error))
            {
                if (exitCode == 0)
                {
                    _logger?.LogWarning($"Git stderr (exit code 0):\n{error.Trim()}");
                }
                else
                {
                    _logger?.LogError($"Git stderr (exit code {exitCode}):\n{error.Trim()}");
                }
            }

            _logger?.LogInformation($"Git command completed with exit code: {exitCode}");

            string rawError = error.Trim();
            string? errorMessage = exitCode != 0
                ? CreateUserFacingErrorMessage(
                    string.IsNullOrWhiteSpace(rawError)
                        ? $"Git command failed with exit code {exitCode}."
                        : rawError,
                    enableLongPathSupport)
                : null;

            return new GitProcessResult
            {
                Success = exitCode == 0,
                ExitCode = exitCode,
                Output = output,
                ErrorMessage = errorMessage
            };
        }
        catch (Exception ex)
        {
            _logger?.LogException(ex, $"Failed to execute Git command: git {displayArgs}");
            string errorMessage = CreateUserFacingErrorMessage(
                $"Failed to execute Git command: {ex.Message}",
                enableLongPathSupport);

            return new GitProcessResult
            {
                Success = false, ExitCode = -1, ErrorMessage = errorMessage
            };
        }
    }

    internal static string BuildGitArguments(string arguments, bool enableLongPathSupport)
    {
        return enableLongPathSupport
            ? $"-c core.longpaths=true {arguments}"
            : arguments;
    }

    internal static List<string> BuildGitArgumentList(IEnumerable<string> arguments, bool enableLongPathSupport)
    {
        var args = arguments.ToList();
        if (enableLongPathSupport)
        {
            args.Insert(0, "core.longpaths=true");
            args.Insert(0, "-c");
        }

        return args;
    }

    internal static string QuoteForDisplay(string arg)
    {
        if (string.IsNullOrEmpty(arg))
        {
            return "\"\"";
        }

        return arg.Any(char.IsWhiteSpace) || arg.Contains('"')
            ? $"\"{arg.Replace("\"", "\\\"")}\""
            : arg;
    }

    /// <summary>
    /// Indicates whether a remove failure is caused by uncommitted changes,
    /// in which case the user can be offered a force remove.
    /// </summary>
    public static bool IsDirtyWorktreeError(string? errorMessage)
    {
        if (string.IsNullOrWhiteSpace(errorMessage))
        {
            return false;
        }

        return errorMessage.Contains("modified or untracked files", StringComparison.OrdinalIgnoreCase) ||
            errorMessage.Contains("contains modified or untracked files", StringComparison.OrdinalIgnoreCase) ||
            errorMessage.Contains("forcing it", StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>
    /// Indicates whether a remove failure is caused by a worktree lock,
    /// in which case the user can be offered a force remove (sent as -f -f).
    /// </summary>
    public static bool IsLockedWorktreeError(string? errorMessage)
    {
        if (string.IsNullOrWhiteSpace(errorMessage))
        {
            return false;
        }

        return errorMessage.Contains("locked working tree", StringComparison.OrdinalIgnoreCase) ||
            errorMessage.Contains("remove -f -f", StringComparison.OrdinalIgnoreCase) ||
            errorMessage.Contains("unlock first", StringComparison.OrdinalIgnoreCase) ||
            (errorMessage.Contains("locked", StringComparison.OrdinalIgnoreCase) &&
                errorMessage.Contains("worktree", StringComparison.OrdinalIgnoreCase));
    }

    /// <summary>
    /// Indicates whether a remove failure should offer the user a force remove dialog
    /// (dirty worktree or locked worktree — force covers both).
    /// </summary>
    public static bool ShouldOfferForceRemove(string? errorMessage) =>
        IsDirtyWorktreeError(errorMessage) || IsLockedWorktreeError(errorMessage);

    internal static bool IsLongPathError(string? errorMessage)
    {
        if (string.IsNullOrWhiteSpace(errorMessage))
        {
            return false;
        }

        return LongPathErrorPatterns.Any(
            pattern => errorMessage.Contains(pattern, StringComparison.OrdinalIgnoreCase));
    }

    internal static string CreateUserFacingErrorMessage(string errorMessage, bool longPathSupportWasEnabled)
    {
        if (!IsLongPathError(errorMessage))
        {
            return errorMessage;
        }

        string supportStatus = longPathSupportWasEnabled
            ? "Git Worktree Manager already ran this Git command with process-scoped core.longpaths=true."
            : "This looks like a Git for Windows long path error.";

        return string.Join(
            Environment.NewLine + Environment.NewLine,
            errorMessage,
            supportStatus +
            " If the same repository also fails from a terminal or another Git tool, enable Git for Windows long paths globally with:",
            "git config --global core.longpaths true",
            "This is a Git for Windows setting, not a Visual Studio setting.");
    }

    /// <inheritdoc />
    public async Task<WorktreeStatus> GetWorktreeStatusAsync(string path, string branch, CancellationToken ct)
    {
        try
        {
            if (string.IsNullOrEmpty(path) || !Directory.Exists(path))
            {
                return new WorktreeStatus(0, 0, 0, 0, false);
            }

            // 1. Check modifications and untracked files
            // Use a shorter timeout for status (e.g., 10s) as we don't want to hang the UI enrichment
            using var statusCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            statusCts.CancelAfter(10000);

            // --porcelain=v1 is stable. 
            // -uno (no untracked) is fast, but user specifically wants untracked.
            // Using -unormal is usually faster than -uall.
            GitProcessResult statusRes =
                await ExecuteGitCommandAsync(path, new[] { "status", "--porcelain=v1", "-unormal" }, statusCts.Token);

            int modified = 0;
            int untracked = 0;

            if (statusRes.Success && !string.IsNullOrWhiteSpace(statusRes.Output))
            {
                string[] lines = statusRes.Output.Split(new[] { '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
                foreach (string line in lines)
                {
                    if (line.Length < 3)
                    {
                        continue;
                    }

                    string status = line.Substring(0, 2);
                    if (status == "??")
                    {
                        untracked++;
                    }
                    else
                    {
                        modified++;
                    }
                }
            }

            // 2. Check Ahead/Behind (Incoming/Outgoing)
            int ahead = 0;
            int behind = 0;
            bool hasUpstream = false;

            try
            {
                // Check if we have an upstream. Use a very short timeout.
                using var upstreamCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
                upstreamCts.CancelAfter(5000);

                GitProcessResult upstreamRes = await ExecuteGitCommandAsync(path,
                    new[] { "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}" }, upstreamCts.Token);

                if (upstreamRes.Success && !string.IsNullOrWhiteSpace(upstreamRes.Output))
                {
                    // We have an upstream tracking branch
                    hasUpstream = true;
                    
                    // Get ahead/behind counts
                    GitProcessResult countRes = await ExecuteGitCommandAsync(path,
                        new[] { "rev-list", "--left-right", "--count", "HEAD...@{u}" }, upstreamCts.Token);
                    if (countRes.Success && !string.IsNullOrWhiteSpace(countRes.Output))
                    {
                        string[] parts = countRes.Output.Trim()
                            .Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                        if (parts.Length >= 2)
                        {
                            if (int.TryParse(parts[0], out int canForward))
                            {
                                ahead = canForward;
                            }

                            if (int.TryParse(parts[1], out int canPull))
                            {
                                behind = canPull;
                            }
                        }
                    }
                }
                // If upstreamRes failed or output is empty, hasUpstream remains false (local-only branch)
            }
            catch (Exception)
            {
                // If upstream check fails, it's a local-only branch
                _logger?.LogWarning($"No upstream found for {path}. Branch is local-only.");
                hasUpstream = false;
            }

            return new WorktreeStatus(modified, untracked, behind, ahead, hasUpstream);
        }
        catch (OperationCanceledException)
        {
            return new WorktreeStatus(0, 0, 0, 0, false);
        }
        catch (Exception ex)
        {
            _logger?.LogException(ex, $"Failed to get status for {path}");
            return new WorktreeStatus(0, 0, 0, 0, false);
        }
    }

    /// <summary>
    /// Attempts to kill a process safely.
    /// </summary>
    private static void TryKillProcess(Process process)
    {
        try
        {
            if (!process.HasExited)
            {
                process.Kill(true);
            }
        }
        catch
        {
            // Ignore errors when killing process
        }
    }

    /// <summary>
    /// Internal result type for Git process execution.
    /// </summary>
    private record GitProcessResult
    {
        public bool Success { get; init; }
        public int ExitCode { get; init; }
        public string? Output { get; init; }
        public string? ErrorMessage { get; init; }
    }
}
